api/views.py

```diff
@@ -49,6 +51,40 @@ class TimerViewSet(viewsets.ModelViewSet):
     serializer_class = serializers.TimerSerializer
     filterset_class = filters.TimerFilter
 
+    def __timer_operation(self, pk, func):
+        try:
+            timer = models.Timer.objects.get(pk=pk)
+            return func(timer)
+        except models.Timer.DoesNotExist:
+            return Response(
+                {"detail": "timer does not exist"},
+                status=status.HTTP_404_NOT_FOUND
+            )
+
+    @action(detail=True, methods=['patch'])
+    def stop(self, request, pk=None):
+        def do_stop(timer):
+            if not timer.active:
+                return Response(
+                    {"detail": "timer already stopped"},
```

Comment:
Ditto here re: `detail` -- I'm not actually inclined to indicate the timer is already stopped. I think just returning a successful response would be fine. That said I could also see a `non_fields_errors` response here.

Reply:
I was brooding a bit about this, but concur.

If this is not intended, it should be intercepted in the model methods (restart/stop) instead. This might cause possible race conditions that a client cannot prevent: If multiple clients trigger "restart" or "stop" at the same time, start/stop dates will be overwritten.

Not a problem for me, I think pretty much for noone. I will leave it to you to fix this if it really is a problem ;-)